### ISSUE TITLE
CLI supports assuming a different role locally

### ISSUE THREAD

**Issue by mrgrain** [aws/aws-cdk#8272]
To add a new command line flag `--assume-role-arn [role-arn]` which will  assume the given role locally before executing any commands. 

This should be semantically equivalent to setting the following configuration when using the aws-sdk:
```
AWS.config.credentials = new AWS.ChainableTemporaryCredentials({
  params: {RoleArn: '...'}
});
```

It would be expected to work with all variations of `--ec2creds`.

### Use Case
Sometimes the current credentials are not enough to execute CDK. This is different from `--role-arn` which happens on ChangeSet execution. However CDK needs to do things before that (e.g. uploading assets).

This is also supported by the aws-cli via profiles, which proves the use case. However working with cli profiles from the JavaScript world is painful (even though support got better) and is more complicated to automate in CI/CD environments.

Terraform has a similar option: https://www.terraform.io/docs/providers/aws/index.html#assume_role

### Proposed Solution

See above. We need a new cli flag. Somewhere in `aws-auth` the value would need to be passed in to `ChainableTemporaryCredentials` which would be used to configure the SDK.

* [x ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

**Comment by richardhboyd** [comment]
I'm curious about the use-case that the existing `--profile` flag doesn't work.

**Comment by mrgrain** [comment]
The existing `--profile` flag does work, but it's non native and imho relatively  complicated way to setup credentials this way.

In order to assume a role using profiles, one has to create a `.aws/config` file with the appropriate contents. This either has to be done manually or it requires installing the `aws-cli` tool. In many CI/CD situations, the default location in `$HOME` could not be used either and `AWS_CONFIG_FILE` would have to be set as well.

CDK already sets a precedent for this with the `--ec2creds` flag. This is also entirely possible with profiles, however the common use case case justifies the simplification.

**tldr; I believe it's a common scenario and it shouldn't be necessary to step outside of CDK to be able to do this.**

**Comment by mrgrain** [comment]
Actually, according to [[master] CDK CLI Authentication Issues](https://github.com/aws/aws-cdk/issues/1656) using profiles does not work to assume roles at all: 

> CDK CLI does not support using EC2/CodeBuild credentials to assume roles in profiles

**Comment by richardhboyd** [comment]
I’m curious bout the use-case for profiles in CodeBuild. Usually a CodeBuild Job has an attached Role that is used to perform the build task, typically if multiple roles are needed (I.e. to interact with multiple target environments) the user would create multiple Jobs with distinct roles to maintain least privilege permissions.

**Comment by mrgrain** [comment]
I cannot speak for CodeBuild, as we are using Jenkins running on an EC2. I can only assume there would be a similar use case to the one we have. In any case, apparently it was warranted that @rix0rrr included the quoted section into the linked master issue. 

For Jenkins/EC2 the issue is that agent instances run on a single shared EC2 with a single shared instance role. In order to maintain least privilege permissions, we have each stage of every Job assume permissions appropriate to what the stage is doing. This could be assuming a role on AWS to deploy (multiple) CDK stacks or upload a file to S3, but also credentials to install private npm packages. Again, I'm not actively using CodeBuild, but this might be equivalent to different CodeBuild phases requiring different roles.

**Comment by danwashusen** [comment]
For anyone else trying to plug CDK into an automated build environment, here's my solution:
```
    - ASSUMED_FOLE=$(aws sts assume-role --role-arn "$AWS_ROLE_ARN" --role-session-name $GITLAB_USER_LOGIN --output json)
    - export AWS_ACCESS_KEY_ID=$(echo $ASSUMED_FOLE | jq -r .Credentials.AccessKeyId)
    - export AWS_SECRET_ACCESS_KEY=$(echo $ASSUMED_FOLE | jq -r .Credentials.SecretAccessKey)
    - export AWS_SESSION_TOKEN=$(echo $ASSUMED_FOLE | jq -r .Credentials.SessionToken)
    - cdk bootstrap
    - cdk deploy foobar
```

It requires the standard `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` to be defined for the initial sts call.

**Comment by mrgrain** [comment]
CDK v2 uses different roles for different tasks, so a single cli option doesn't really make much sense. This can all be achieved using credential plugins and/or custom synthesizers.
